staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go

```diff
@@ -254,13 +256,24 @@ func ResetObjectMetaForStatus(meta, existingMeta Object) {
 }
 
 // MarshalJSON implements json.Marshaler
+// MarshalJSON may get called on pointers or values, so implement MarshalJSON on value.
+// http://stackoverflow.com/questions/21390979/custom-marshaljson-never-gets-called-in-go
 func (f Fields) MarshalJSON() ([]byte, error) {
-	return json.Marshal(&f.Map)
+	if f.Raw == nil {
+		return []byte("null"), nil
+	}
+	return f.Raw, nil
 }
 
 // UnmarshalJSON implements json.Unmarshaler
 func (f *Fields) UnmarshalJSON(b []byte) error {
-	return json.Unmarshal(b, &f.Map)
+	if f == nil {
+		return errors.New("metav1.Fields: UnmarshalJSON on nil pointer")
+	}
+	if !bytes.Equal(b, []byte("null")) {
+		f.Raw = append(f.Raw[0:0], b...)
```

Comment:
does this allow storing non-map values `1`, `true`, `[""]`, `""`, etc? what about map values that aren't schema-compatible with the old `map[string]Field` structure? what ensures the raw JSON persisted in this field doesn't break old clients?

Reply:
old clients as in, clients written while it was an alpha api?

Reply:
yeah, we can't ever send data that one of those clients would fail to deserialize

Reply:
we could validate that it is just a map of maps, but would this require us to deserialize it on every update?

Reply:
possibly, but I expected format validation on update anyway...

Reply:
Actually this is already validated when we make sure it will convert into an smd fieldset

Reply:
and that happens on every write (create/update/all patch types/subresources/etc)?

Reply:
do we have negative tests ensuring malformed data submitted by clients is rejected?